### PR TITLE
[mlir][tensor] Restrict the verifier for tensor.pack/tensor.unpack

### PR DESCRIPTION
**Pull request by banach-space**
Restricts the verifier for tensor.pack and tensor.unpack Ops so that the
following is no longer allowed:

```mlir
  %c8 = arith.constant 8 : index
  %0 = tensor.pack %input inner_dims_pos = [0, 1] inner_tiles = [8, %c8] into %output : tensor<?x?xf32> -> tensor<?x?x8x8xf32>
```

Specifically, in line with other Tensor Ops, require:
  * a dynamic dimensions for each (dynamic) SSA value,
  * a static dimension for each static size (attribute).

In the example above, a static dimension (8) is mixed with a dynamic
size (%c8).

Note that this is mostly deleting existing code - that's because this
change simplifies the logic in verifier.

For more context:
  * https://discourse.llvm.org/t/tensor-ops-with-dynamic-sizes-which-behaviour-is-more-correct

### CODE REVIEW

File: mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h [hunk 1]
```diff
@@ -261,9 +260,9 @@ struct BufferizationOptions {
   using AnalysisStateInitFn = std::function<void(AnalysisState &)>;
   /// Tensor -> MemRef type converter.
   /// Parameters: Value, memory space, func op, bufferization options
-  using FunctionArgTypeConverterFn = std::function<BaseMemRefType(
-      TensorType, Attribute memorySpace, FunctionOpInterface,
-      const BufferizationOptions &)>;
+  using FunctionArgTypeConverterFn =
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h:263]: It looks like there are some unrelated bufferization changes in this PR.

Comment (banach-space) [reply mlir/include/mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h]: Sorry about that, wrong base branch 🤦🏻 This is now fixed.

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 2]
```diff
@@ -3865,22 +3865,15 @@ static LogicalResult commonVerifierPackAndUnPackOp(OpTy packOrUnPack) {
...
-            std::optional<int64_t> constTileSize =
-                getConstantIntValue(std::get<1>(it));
             int64_t shape = std::get<0>(it);
-            if (!constTileSize) {
-              // If specified tile size is dynamic, output shape should
-              // be dynamic too.
-              return ShapedType::isDynamic(shape);
+            if (Attribute attr =
+                    llvm::dyn_cast_if_present<Attribute>(std::get<1>(it))) {
+              if (IntegerAttr intAttr = dyn_cast_or_null<IntegerAttr>(attr)) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:3871]: I think you can assume that an attribute is always an `IntegerAttr`. That should be verified by the auto-generated verifier.

**Review by matthias-springer — APPROVED** [review]
